File: Leetcode/Leetcode-DSA-course/Hashing/2428-equal-row-and-column-pairs/equal-row-and-column-pairs.ts
function equalPairs(grid: number[][]): number {
    function convertToKey(arr: number[]) {
        let str = "";
        for (let num of arr)
            str += num + ","
        return str;
    }

    const dic = new Map();
    for (let row of grid) {
        let key = convertToKey(row)
        dic.set(key, (dic.get(key) || 0) + 1)
    }
    const dic2 = new Map()
    for (let col = 0; col < grid[0].length; col++) {
        let currentCol = []
        for (let row = 0; row < grid.length; row++) {
            currentCol.push(grid[row][col])
        }
        let key = convertToKey(currentCol)
        dic2.set(key, (dic2.get(key) || 0) + 1)
    }

    let ans = 0;
    for (let [arr, val] of dic) {
        ans += val * (dic2.get(arr) || 0);
    }
    return ans;

};

File: Leetcode/Leetcode-DSA-course/Hashing/2428-equal-row-and-column-pairs/equal-row-and-column-pairs.java
class Solution {
    public int equalPairs(int[][] grid) {
        Map<String, Integer> dic = new HashMap<>();
        for (int[] row : grid) {
            String key = convertToKey(row);
            dic.put(key, dic.getOrDefault(key, 0) + 1);
        }
        Map<String, Integer> dic2 = new HashMap<>();
        for (int col = 0; col < grid[0].length; col++) {
            int[] currentCol = new int[grid.length];
            for (int row = 0; row < grid.length; row++) {
                currentCol[row] = grid[row][col];
            }
            String key = convertToKey(currentCol);
            dic2.put(key, dic2.getOrDefault(key, 0) + 1);
        }

        int ans = 0;
        for (String arr : dic.keySet()) {
            ans += dic.get(arr) * dic2.getOrDefault(arr, 0);
        }
        return ans;
    }

    public String convertToKey(int[] arr) {
        StringBuilder sb = new StringBuilder();
        for (int num : arr) {
            sb.append(num);
            sb.append(",");
        }
        return sb.toString();
    }
}